shell/platform/android/test/io/flutter/embedding/android/FlutterFragmentTest.java

```diff
@@ -108,4 +120,42 @@ public void itCanBeDetachedFromTheEngineAndStopSendingFurtherEvents() {
     verify(mockDelegate, times(1)).onDestroyView();
     verify(mockDelegate, times(1)).onDetach();
   }
+
+  @Test
+  public void itDoesNotRecurseSystemNavigatorPopWhenUsingOnBackPressedCallback() {
+    FlutterJNI mockFlutterJni = mock(FlutterJNI.class);
+    when(mockFlutterJni.isAttached()).thenReturn(true);
+    FlutterEngine engine =
+        new FlutterEngine(
+            RuntimeEnvironment.application, mock(FlutterLoader.class), mockFlutterJni);
+    NavigationChannel navigationChannel = spy(engine.getNavigationChannel());
+    FlutterEngineCache.getInstance().put("my_cached_engine", engine);
+
+    FlutterFragment fragment = FlutterFragment.withCachedEngine("my_cached_engine").build();
+    FragmentActivity activity = spy(Robolectric.setupActivity(FragmentActivity.class));
+    activity
+        .getSupportFragmentManager()
+        .beginTransaction()
+        .replace(android.R.id.content, fragment)
+        .commitNow();
+    engine.getPlatformChannel().channel.invokeMethod("SystemNavigator.pop", null);
+    verify(navigationChannel, never()).popRoute();
```

Comment:
Why we shold test `popRoute()`? there is no relation between `SystemNavigator.pop` and `NavigationChannel`

Reply:
This test unfortunately doesn't work for reasons I haven't had time to resolve. I can't figure out why the `invokeMethod` call isn't triggering the `PlatformPlugin` the way I'd expect. Anyways, the reason is that your other change would cause that specific line to fail. Your change would introduce the following approximate call stack:

```
at NavigationChannel#popRoute
at FlutterFragment$OnBackPressedCallback#handleOnBackPressed
at OnBackPressedDispatcher#onBackPressed
at PlatformPlugin#popSystemNavigation
```

This stack would then recurse on the dart side to call `popSystemNavigation` again, creating a serious bug. I verified in my locally built app that this is exactly what happens. Reproducing it in a test requires a lot of moving parts, but this test case at a high level (if it worked) should cover the core behavior.

Reply:
So if I understand correctly you call the `onBackPressed()` function (from Flutter Fragment) in `OnBackPressedCallback`not in :

```
@Override
public void onBackPressed() {
    ----------
}
```

Reply:
Yes, that is the core concept of my concern.

Reply:
i think this will be the best fix for all the use case. wdyt?
```
public void popSystemNavigator() {

  if (!(activity instanceof FragmentActivity)) {
    activity.finish();
    return;
  }

  FragmentManager fragmentManager = ((FragmentActivity)) activity).getSupportFragmentManager();

  if (fragmentManager.getBackStackEntryCount() > 0) {
    fragmentManager.popBackStack();
  } else {
    activity.finish();
  }
}
```

Reply:
That still doesn't address the possibility that other fragments might register OnBackPressedCallbacks that should be triggered. The only correct solution is to disable the FlutterFragment's own callback.

Reply:
We must fix the issue without breaking change .

Reply:
There is a solution that isn't a breaking change, it's just a lot more work. That said, some clients besides myself might be relying on the existing broken behavior, which technically means that even your change in its current form is a breaking change.

Reply:
Maybe you can create new PR with the solution . It will be great.

Reply:
I've already talked extensively with @xster about it, but I don't have the time to do it myself right now. It also involves some complex work on the Dart side of things that I wouldn't trust myself with. I mostly just know Android.